mule-extensions-api/src/main/java/org/mule/runtime/extension/internal/loader/enricher/BackPressureDeclarationEnricher.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.extension.internal.loader.enricher;
+
+import static java.lang.String.format;
+import static org.mule.metadata.api.model.MetadataFormat.JAVA;
+import static org.mule.runtime.api.meta.ExpressionSupport.NOT_SUPPORTED;
+import static org.mule.runtime.api.meta.model.parameter.ParameterGroupModel.DEFAULT_GROUP_NAME;
+import static org.mule.runtime.extension.api.ExtensionConstants.BACK_PRESSURE_STRATEGY_PARAMETER_DESCRIPTION;
+import static org.mule.runtime.extension.api.ExtensionConstants.BACK_PRESSURE_STRATEGY_PARAMETER_NAME;
+import static org.mule.runtime.extension.api.annotation.param.display.Placement.ADVANCED_TAB;
+import static org.mule.runtime.extension.api.loader.DeclarationEnricherPhase.STRUCTURE;
+
+import org.mule.metadata.api.annotation.EnumAnnotation;
+import org.mule.metadata.api.builder.BaseTypeBuilder;
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.metadata.java.api.annotation.ClassInformationAnnotation;
+import org.mule.runtime.api.meta.model.declaration.fluent.ExtensionDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ParameterDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.SourceDeclaration;
+import org.mule.runtime.api.meta.model.display.LayoutModel;
+import org.mule.runtime.extension.api.declaration.fluent.util.IdempotentDeclarationWalker;
+import org.mule.runtime.extension.api.loader.DeclarationEnricher;
+import org.mule.runtime.extension.api.loader.DeclarationEnricherPhase;
+import org.mule.runtime.extension.api.loader.ExtensionLoadingContext;
+import org.mule.runtime.extension.api.runtime.source.BackPressureMode;
+import org.mule.runtime.extension.internal.property.BackPressureStrategyModelProperty;
+
+import java.util.Optional;
+
+/**
+ * Adds a parameter for back pressure on message sources that apply
+ *
+ * @since 1.1
+ */
+public class BackPressureDeclarationEnricher implements DeclarationEnricher {
+
+  @Override
+  public void enrich(ExtensionLoadingContext extensionLoadingContext) {
+    final ExtensionDeclaration extensionDeclaration = extensionLoadingContext.getExtensionDeclarer().getDeclaration();
+    new IdempotentDeclarationWalker() {
+
+      @Override
+      protected void onSource(SourceDeclaration sourceDeclaration) {
+        Optional<BackPressureStrategyModelProperty> backPressureStrategyModelProperty =
```

Comment:
there's a big design problem here. You have a core enricher implementing a language rule based on a model property which is not API. So this depends on a third party to provide something that's not API, in order to do something that's a core language rule.

If you leave this as is, that model property should be API. But that MP shouldn't be API. Maybe leave it for another iteration, but this is clearly pointing towards a design issue.

Reply:
I get your point, the idea was to parse the relevant information for the enrichment in the loading process, and have the enrichment done here. Do you see other way to achieve this?

Reply:
for now let's just make the Model Property API and make it explicit in the SourceModelParser (not just as an additional model property). We can revisit this later.